### PR TITLE
Fix row focus handling for DetailsList

### LINKED ISSUE #2051: DetailsList: Resizing columns doesn't work
### Bug Report

- __Package version(s)__: latest

#### Priorities and help requested (not applicable if asking question):

Requested priority: High

#### Describe the issue:

Try to resize columns here:
https://codepen.io/dzearing/pen/XRvXoL

### PR DESCRIPTION
**Pull request by ThomasMichon**
This change fixes some issues impacting the `DetailsList` and keyboard focus:
1. Hitting up when focused on the top row was not moving focus to the header.
1. Reloading the list was not automatically setting focus to the first row.
1. Resizing columns was causing the entire header to re-render, disrupting the operation.

This also fixes #2051 by adjusting how the overrides for rendering are handled.

### CODE REVIEW

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.tsx [hunk 1]
```diff
@@ -246,11 +255,7 @@ export class DetailsList extends BaseComponent<IDetailsListProps, IDetailsListSt
       selectAllVisibility = SelectAllVisibility.none;
     }
 
-    const {
-      onRenderDetailsHeader = this._onRenderDetailsHeader
-    } = this.props;
-
-    const DetailsHeader = (props: IDetailsHeaderProps) => onRenderDetailsHeader(props, this._onRenderDetailsHeader);
+    const DetailsHeader = this._DetailsHeader;
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.tsx]: confused on this._DetailsHeader...

Comment (ThomasMichon) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.tsx]: I'm trying to take advantage of `<Function>` syntax to keep the markup below clean. For this to work I need to store that function as a field. We already have `_onRenderDetailsHeader` as its own member, so I need to give the final function a new name.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.tsx [hunk 2]
```diff
@@ -273,7 +278,7 @@ export class DetailsList extends BaseComponent<IDetailsListProps, IDetailsListSt
           <div onKeyDown={ this._onHeaderKeyDown } role='presentation'>
             { isHeaderVisible && (
               <DetailsHeader
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.tsx]: This is probably good enough, no need for the funky private:

```tsx
onRenderDetailsHeader({
  props
}, this._onRenderDetailsHeader);
```

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRow.tsx [hunk 3]
```diff
@@ -320,6 +325,18 @@ export class DetailsRow extends BaseComponent<IDetailsRowProps, IDetailsRowState
     selection.toggleIndexSelected(this.props.itemIndex);
   }
 
+  private _onRootRef(root: FocusZone) {
+    const {
+      refs
+    } = this;
+
+    if (root) {
+      refs.root = ReactDOM.findDOMNode(root) as HTMLElement;
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsRow.tsx]: ... is there a good reason for this over using resolveRef? (there might be, not sure.) 

Comments might be good to explain what's up here.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.tsx [hunk 4]
```diff
@@ -271,27 +271,25 @@ export class DetailsList extends BaseComponent<IDetailsListProps, IDetailsListSt
...
+              onColumnIsSizingChanged: this._onColumnIsSizingChanged,
+              onColumnAutoResized: this._onColumnAutoResized,
+              groupNestingDepth: groupNestingDepth,
+              isAllCollapsed: isCollapsed,
+              onToggleCollapseAll: this._onToggleCollapse,
+              ariaLabel: ariaLabelForListHeader,
+              ariaLabelForSelectAllCheckbox: ariaLabelForSelectAllCheckbox,
+              ariaLabelForSelectionColumn: ariaLabelForSelectionColumn,
+              selectAllVisibility: selectAllVisibility
+            }) }
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.tsx]: you need to pass in this._onRenderDetailsHeader as the defaultRender parameter I believe.

Comment (ThomasMichon) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.tsx]: Hm, good point. I could just pass it directly and save myself a redefinition of the function. As-written the value does get passed.

**Pull request merged**